x-pack/plugins/security_solution/server/lib/risk_engine/utils/create_ds.ts

```diff
@@ -0,0 +1,214 @@
+      { logger }
+    );
+
+    return;
+  } catch (err) {
+    logger.error(`Failed to PUT mapping for alias ${alias}: ${err.message}`);
+    throw err;
+  }
+};
+/**
+ * Updates the underlying mapping for any existing concrete indices
+ */
+const updateIndexMappings = async ({
+  logger,
+  esClient,
+  totalFieldsLimit,
+  concreteIndices,
+}: UpdateIndexMappingsOpts) => {
+  logger.debug(`Updating underlying mappings for ${concreteIndices.length} indices.`);
+
+  // Update total field limit setting of found indices
+  // Other index setting changes are not updated at this time
+  await Promise.all(
+    concreteIndices.map((index) =>
+      updateTotalFieldLimitSetting({ logger, esClient, totalFieldsLimit, concreteIndexInfo: index })
+    )
+  );
+
+  // Update mappings of the found indices.
+  await Promise.all(
+    concreteIndices.map((index) =>
+      updateUnderlyingMapping({ logger, esClient, totalFieldsLimit, concreteIndexInfo: index })
+    )
+  );
+};
+
+interface CreateConcreteWriteIndexOpts {
+  logger: Logger;
+  esClient: ElasticsearchClient;
+  totalFieldsLimit: number;
+  indexPatterns: IIndexPatternString;
+}
+/**
+ * Installs index template that uses installed component template
+ * Prior to installation, simulates the installation to check for possible
+ * conflicts. Simulate should return an empty mapping if a template
+ * conflicts with an already installed template.
+ */
+export const createConcreteWriteIndex = async ({
+  logger,
+  esClient,
+  indexPatterns,
+  totalFieldsLimit,
+}: CreateConcreteWriteIndexOpts) => {
+  logger.info(`Creating data stream - ${indexPatterns.alias}`);
+
+  // check if a concrete write index already exists
+  let concreteIndices: ConcreteIndexInfo[] = [];
+  try {
+    // Specify both the index pattern for the backing indices and their aliases
+    // The alias prevents the request from finding other namespaces that could match the -* pattern
+    const response = await retryTransientEsErrors(
+      () => esClient.indices.getDataStream({ name: indexPatterns.alias, expand_wildcards: 'all' }),
+      { logger }
+    );
+
+    concreteIndices = response.data_streams.map((dataStream) => ({
+      index: dataStream.name,
+      alias: dataStream.name,
+      isWriteIndex: true,
+    }));
+
+    logger.debug(
+      `Found ${concreteIndices.length} concrete indices for ${
+        indexPatterns.alias
+      } - ${JSON.stringify(concreteIndices)}`
+    );
+  } catch (error) {
+    // 404 is expected if no concrete write indices have been created
+    if (error.statusCode !== 404) {
+      logger.error(
+        `Error fetching concrete indices for ${indexPatterns.alias} pattern - ${error.message}`
+      );
+      throw error;
+    }
+  }
+
+  // let concreteWriteIndicesExist = false;
+  const concreteWriteIndicesExist = concreteIndices.length > 0;
+
+  // if a concrete write ds already exists, update the underlying mapping
```

Comment:
Do you mean "datastream write index"? Is it possible for a datastream to exist without a backing index?

I feel like a lot of the terminology in this file is confused as we're dealing with datastreams and not indices; we might need another pass here to make this more accurate and distinct from the alerting utilities.